utils/pattern/src/pattern.rs

```diff
@@ -0,0 +1,85 @@
+use crate::parser::error::ParserError;
+use crate::token::PatternToken;
+use std::{fmt::Debug, str::FromStr};
+
+/// Input trait for a `Interpolator`.
+///
+/// This trait is automatically implemented on a [`Vec<PatternToken>>`][`Vec`].
+///
+/// # Examples
+/// ```
+/// use icu_pattern::{
+///     PatternToken,
+///     Interpolator,
+///     pattern::IntoIterVec,
+/// };
+///
+/// #[derive(Debug, PartialEq)]
+/// enum Element<'s> {
+///     Literal(&'s str),
+///     Token,
+/// }
+///
+/// impl<'s> From<&'s str> for Element<'s> {
+///     fn from(input: &'s str) -> Self {
+///         Self::Literal(input)
+///     }
+/// }
+///
+/// let pattern: IntoIterVec<_> = vec![
+///     PatternToken::Placeholder(0usize),
+///     PatternToken::Literal { content: ", ", quoted: false },
+///     PatternToken::Placeholder(1),
+/// ].into();
+///
+/// let mut interpolator = Interpolator::new(pattern, vec![
+///     vec![
+///         Element::Token
+///     ],
+///     vec![
+///         Element::Token
+///     ],
+/// ]);
+///
+/// assert_eq!(Ok(Some(Element::Token)), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Literal(", "))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Token)), interpolator.try_next());
+/// ```
+pub trait PatternIterator<'p, P> {
+    fn try_next(&mut self) -> Result<Option<PatternToken<'p, P>>, ParserError<<P as FromStr>::Err>>
+    where
+        P: FromStr,
+        P::Err: Debug;
+}
+
+/// An iterator constructed from a [`Vec<PatternToken>>`][`Vec`]
+/// which implements [`PatternIterator`] trait.
+pub struct IntoIterVec<'p, P> {
+    buf: <Vec<PatternToken<'p, P>> as IntoIterator>::IntoIter,
+}
+
+impl<'p, P> Iterator for IntoIterVec<'p, P> {
+    type Item = PatternToken<'p, P>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.buf.next()
+    }
+}
+
+impl<'p, P> From<Vec<PatternToken<'p, P>>> for IntoIterVec<'p, P> {
```

Comment:
Issue: It's sad that you have to consume the `Vec<PatternToken>` for this.  I would rather point to a borrowed, immutable `Vec<PatternToken>`, like you might have from a DataProvider.

Can you make it such that the iterator produces `Option<&'p PatternToken<'p, P>>` instead of `Option<PatternToken<'p, P>>`?

Reply:
That's a good point! I'll try!

Reply:
I failed. I think it's worth doing, but I'm not sure how to achieve it. The issue is that while data provider provided tokens should be references, Parser provided replacements should be owned since the parser will generate the tokens.

I'm not sure if we can achieve it without GAT or even HKT (I hope GAT is enough but it's hard to reason about). I'll keep this issue open because for a perfect behavior we'd like it to operate the way you suggested.

I'll also wait for @Manishearth review and maybe he will help us model it in a way that allows for both behaviors here.

I was inclined to (for ergonomic reasons) try to have Interpolator take `P: Into<PatternIterator>` so that people don't have to `into()` their `Vec` before passing it, but that's exactly what GAT/HKT is needed for, and then in the future Rust I'd like to write it as `P: Into<PatternIterator<Item = Borrow<PatternToken>>` but that's GAT/HKT + Fallible Iterators.

Reply:
I decided to move away from streaming pattern based on the feedback and because I want to prioritize pre-parsed pattern interpolation.